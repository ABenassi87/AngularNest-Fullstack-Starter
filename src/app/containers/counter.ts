export interface CounterState {
  counter: number;
}

export const INITIAL_COUNTER_STATE: CounterState = {
  counter: 0
};

export const INCREMENT = 'INCREMENT';
export const DECREMENT = 'DECREMENT';
export const RESET = 'RESET';
export const INCREMENT_IF_ODD = 'INCREMENT_IF_ODD';
export const DECREMENT_IF_EVEN = 'DECREMENT_IF_EVEN';

export const increment = () => ({ type: INCREMENT });
export const decrement = () => ({ type: DECREMENT });
export const reset = () => ({ type: RESET });
export const incrementIfOdd = () => ({ type: INCREMENT_IF_ODD });
export const decrementIfEven = () => ({ type: DECREMENT_IF_EVEN });

import { Action } from '@ngrx/store';

export const counterReducer = (state = INITIAL_COUNTER_STATE, action: Action) => {
  const { type, payload } = action;

  switch (type) {
    case INCREMENT:
      return { ...state, counter: state.counter + 1 };
      // return Object.assign({}, state, { counter: state.counter + payload.value });
    case DECREMENT:
      return { ...state, counter: state.counter - 1 };
      // return Object.assign({}, state, { counter: state.counter - payload.value });
    case RESET:
      return INITIAL_COUNTER_STATE;
      // return Object.assign({}, state, { counter: 0 });
    default:
      return state;
  }
};

import { Injectable } from '@angular/core';
import { Store } from '@ngrx/store';
import { Actions, Effect } from '@ngrx/effects';

import 'rxjs/add/operator/withLatestFrom';
import 'rxjs/add/operator/filter';
import 'rxjs/add/operator/map';

@Injectable()
export class CounterEffects {
  @Effect() incrementIfOdd$ = this.actions$
    .ofType(INCREMENT_IF_ODD)
    .withLatestFrom(this.state$)
    .filter(([action, state]) => state.counter % 2 === 1)
    .map(increment);

  @Effect() decrementIfEven$ = this.actions$
    .ofType(DECREMENT_IF_EVEN)
    .withLatestFrom(this.state$)
    .filter(([action, state]) => state.counter % 2 === 0)
    .map(decrement);

  constructor(private actions$: Actions, private state$: Store<CounterState>) { }
}
