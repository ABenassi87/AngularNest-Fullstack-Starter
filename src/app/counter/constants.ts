// import { Counter } from './model';
//
// export const INITIAL: Counter = { value: 0 };
//
// export const INCREMENT = '[Counter] INCREMENT';
// export const DECREMENT = '[Counter] DECREMENT';
//
// export const INCREMENT_ASYNC = '[Counter] INCREMENT_ASYNC';
// export const INCREMENT_IF_ODD = '[Counter] INCREMENT_IF_ODD';
