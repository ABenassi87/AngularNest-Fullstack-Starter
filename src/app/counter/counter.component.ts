import { Component } from '@angular/core';
import { Observable } from 'rxjs/Observable';
import { Action, Store } from '@ngrx/store';

import { Counter } from './model';
import { onIncrement, onDecrement } from './actions';

@Component({
  selector: 'app-counter',
  template: `
    <div class="counter">
      <button (click)="onIncrement()">Increment</button>
      <button (click)="onDecrement()">Decrement</button>
      <button (click)="onIncrementAsync()">Increment (Async)</button>
      <button (click)="onIncrementIfOdd()">Increment (If Odd)</button>
      <h3>{{ counter$ }}</h3>
    </div>
  `
})
export class CounterComponent {
  public counter$: Observable<number>;

  constructor(private store: Store<Counter>) {
    this.counter$ = store.select('value');
    console.log(this.counter$)
  }

  onIncrement(): void {
    this.store.dispatch(onIncrement());
  }

  onDecrement(): void {
    this.store.dispatch(onDecrement());
  }

  onIncrementAsync(): void {
    // this.store.dispatch(onIncrementAsync());
  }

  onIncrementIfOdd(): void {
    // this.store.dispatch(incrementIfOdd());
  }
}
