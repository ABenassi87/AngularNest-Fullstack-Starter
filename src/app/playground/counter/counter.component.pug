div(*mobxAutorun='{ dontDetach: true }')
  p Clicked: {{ counter.value }} times, value is {{ counter.evenOrOdd }}.

  div
    button(mat-raised-button (click)='counter.increment()') Increment
    button(mat-raised-button (click)='counter.decrement()') Decrement

  div
    button(mat-raised-button (click)='counter.incrementAsync()') Increment (Async)
    button(mat-raised-button (click)='counter.decrementAsync()') Decrement (Async)
    button(mat-raised-button (click)='counter.incrementIfOdd()') Increment (If Odd)
    button(mat-raised-button (click)='counter.decrementIfEven()') Decrement (If Even)
