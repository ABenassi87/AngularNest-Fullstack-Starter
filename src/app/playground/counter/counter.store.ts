import 'rxjs/add/observable/of';
import 'rxjs/add/operator/delay';
import 'rxjs/add/operator/filter';

import { Injectable } from '@angular/core';
import { observable, action, computed } from 'mobx';
import { Observable } from 'rxjs/Observable';

@Injectable()
export class CounterStore {
  @observable value = 0;

  @action
  increment(): void {
    this.value += 1;
  }

  @action
  decrement(): void {
    this.value -= 1;
  }

  @action
  incrementAsync(): void {
    setTimeout(() => this.increment(), 1000);
  }

  @action
  decrementAsync(): void {
    Observable.of(null)
      .delay(1000)
      .subscribe(() => this.decrement());
  }

  @action
  incrementIfOdd(): void {
    if (Math.abs(this.value % 2) === 1) {
      this.increment();
    }
  }

  @action
  decrementIfEven(): void {
    Observable.of(null)
      .filter(() => this.value % 2 === 0)
      .subscribe(() => this.decrement());
  }

  @action
  startCount() {

  }

  @action
  cancelCount() {

  }

  @computed
  get evenOrOdd(): string {
    return this.value % 2 === 0 ? 'even' : 'odd';
  }
}
