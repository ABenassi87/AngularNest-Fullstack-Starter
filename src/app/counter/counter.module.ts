import { NgModule } from '@angular/core';
import { Routes, RouterModule } from '@angular/router';
import { MatButtonModule } from '@angular/material';

import { CounterComponent } from './counter.component';

export const ROUTES: Routes = [
  { path: '', component: CounterComponent }
];

@NgModule({
  imports: [
    MatButtonModule,
    RouterModule.forChild(ROUTES)
  ],
  declarations: [
    CounterComponent
  ]
})
export class CounterModule {}
