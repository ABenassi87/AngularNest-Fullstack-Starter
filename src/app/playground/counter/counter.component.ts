import { Component } from '@angular/core';

import { CounterStore } from './counter.store';

@Component({
  selector: 'playground--counter',
  template: `
    <div *mobxAutorun="{ dontDetach: true }">
      <p>Clicked: {{ counter.value }} times, value is {{ counter.evenOrOdd }}.</p>
      <button mat-raised-button (click)="counter.increment()">Increment</button>
      <button mat-raised-button (click)="counter.decrement()">Decrement</button>
      <button mat-raised-button (click)="counter.incrementAsync()">Increment (Async)</button>
      <button mat-raised-button (click)="counter.decrementAsync()">Decrement (Async)</button>
      <button mat-raised-button (click)="counter.incrementIfOdd()">Increment (If Odd)</button>
      <button mat-raised-button (click)="counter.decrementIfEven()">Decrement (If Even)</button>
    </div>
  `,
  styles: [``]
})
export class CounterComponent {
  constructor(public counter: CounterStore) {}
}
