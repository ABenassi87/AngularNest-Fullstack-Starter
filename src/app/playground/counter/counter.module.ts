import { NgModule } from '@angular/core';
import { Routes, RouterModule } from '@angular/router';
import { MatButtonModule } from '@angular/material';
import { MobxAngularModule } from 'mobx-angular';

import { CounterComponent } from './counter.component';
import { CounterStore } from './counter.store';

const ROUTES: Routes = [
  { path: '', component: CounterComponent }
];

@NgModule({
  imports: [
    MobxAngularModule,
    MatButtonModule,
    RouterModule.forChild(ROUTES)
  ],
  declarations: [
    CounterComponent
  ],
  providers: [
    CounterStore
  ]
})
export class CounterModule {}
