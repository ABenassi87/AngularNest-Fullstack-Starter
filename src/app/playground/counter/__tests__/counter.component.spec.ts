import { ComponentFixture, TestBed } from '@angular/core/testing';
import { By } from '@angular/platform-browser';
import { DebugElement } from '@angular/core';
import { MobxAutorunDirective } from 'mobx-angular';

import { CounterComponent } from '../counter.component';
import { CounterStore } from '../counter.store';

describe('CounterComponent', () => {
  let fixture: ComponentFixture<CounterComponent>;

  beforeEach(() => {
    TestBed.configureTestingModule({
      declarations: [
        MobxAutorunDirective,
        CounterComponent
      ],
      providers: [CounterStore]
    });

    fixture = TestBed.createComponent(CounterComponent);
    fixture.detectChanges();
  });

  it('renders markup to snapshot', () => {
    expect(fixture).toMatchSnapshot();
  });
});
