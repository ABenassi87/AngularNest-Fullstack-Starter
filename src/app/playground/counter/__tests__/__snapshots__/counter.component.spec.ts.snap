// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CounterComponent renders markup to snapshot 1`] = `
<playground--counter
  counter={[Function CounterStore]}
>
  <div>
    <p>
      Clicked: 0 times, value is even.
    </p>
    <div>
      <button
        mat-raised-button="mat-raised-button"
      >
        Increment
      </button>
      <button
        mat-raised-button="mat-raised-button"
      >
        Decrement
      </button>
    </div>
    <div>
      <button
        mat-raised-button="mat-raised-button"
      >
        Increment (Async)
      </button>
      <button
        mat-raised-button="mat-raised-button"
      >
        Decrement (Async)
      </button>
      <button
        mat-raised-button="mat-raised-button"
      >
        Increment (If Odd)
      </button>
      <button
        mat-raised-button="mat-raised-button"
      >
        Decrement (If Even)
      </button>
    </div>
  </div>
</playground--counter>
`;
